src/wpantund/Pcap.cpp

```diff
@@ -0,0 +1,375 @@
+PcapManager::insert_fd(int fd)
+{
+	int ret = -1;
+	int save_errno;
+	int set = 1;
+	PcapGlobalHeader header;
+
+	// Prepare the PCAP header.
+	header.mMagic = PCAP_MAGIC;
+	header.mVerMaj = PCAP_VERSION_MAJOR;
+	header.mVerMin = PCAP_VERSION_MINOR;
+	header.mGMTOffset = 0;
+	header.mAccuracy = 0;
+	header.mSnapshotLengthField = PCAP_PACKET_MAX_SIZE;
+	header.mDLT = PCAP_DLT_PPI;
+
+#ifdef SO_NOSIGPIPE
+	setsockopt(fd, SOL_SOCKET, SO_NOSIGPIPE, (void *)&set, sizeof(int));
+#endif
+
+	// Send the PCAP header.
+	ret = static_cast<int>(write(fd, &header, sizeof(header)));
+
+	if (ret < 0) {
+		save_errno = errno;
+		syslog(LOG_ERR, "PcapManager::insert_fd: Call to send() on fd %d failed: %s (%d)", fd, strerror(errno), errno);
+		goto bail;
+	}
+
+	mFDSet.insert(fd);
+
+	ret = 0;
+
+bail:
+	if (ret < 0) {
+		errno = save_errno;
+	}
+
+	return ret;
+}
+
+int
+PcapManager::new_fd(void)
+{
+	int ret = -1;
+	int save_errno;
+	int fd[2] = { -1, -1 };
+
+	ret = socketpair(PF_UNIX, SOCK_DGRAM, 0, fd);
+
+	if (ret < 0) {
+		save_errno = errno;
+		syslog(LOG_ERR, "PcapManager::new_fd: Call to socketpair() failed: %s (%d)", strerror(errno), errno);
+		goto bail;
+	}
+
+	ret = insert_fd(fd[1]);
+
+	if (ret < 0) {
+		save_errno = errno;
+		goto bail;
+	}
+
+	ret = fd[0];
+
+bail:
+
+	if (ret < 0) {
+		close(fd[0]);
+		close(fd[1]);
+
+		errno = save_errno;
+	}
+
+	return ret;
+}
+
+void
+PcapManager::close_fd_set(const std::set<int>& x)
+{
+	if (&x == &mFDSet) {
+		// Special case where we are asked to close everything.
+		std::set<int> copy(x);
+
+		close_fd_set(copy);
+	} else {
+		std::set<int>::const_iterator iter;
+
+		for ( iter  = x.begin()
+			; iter != x.end()
+			; ++iter
+		) {
+			const int fd = *iter;
+			close(fd);
+			mFDSet.erase(fd);
+		}
+	}
+}
+
+void
+PcapManager::push_packet(const PcapPacket& packet)
+{
+	std::set<int>::const_iterator iter;
+	std::set<int> remove_set;
+
+	require_noerr(packet.get_status(), bail);
+
+	for ( iter  = mFDSet.begin()
+	    ; iter != mFDSet.end()
+		; ++iter
+	) {
+		int ret;
+
+		// Send the PCAP frame.
+		ret = static_cast<int>(write(
+			*iter,
+			packet.get_data_ptr(),
+			packet.get_data_len()
+		));
+
+		__ASSERT_MACROS_check(ret >= 0);
+
+		if (ret < 0) {
+			// Since we can't remove this file descriptor
+			// from the set while we are iterating through it,
+			// we add it to the remove set for later removal.
+			remove_set.insert(*iter);
+		}
+	}
+
+	close_fd_set(remove_set);
+
+bail:
+	return;
+}
+
+int
+PcapManager::update_fd_set(fd_set *read_fd_set, fd_set *write_fd_set, fd_set *error_fd_set, int *max_fd, cms_t *timeout)
+{
+	std::set<int>::const_iterator iter;
+
+	for ( iter  = mFDSet.begin()
+	    ; iter != mFDSet.end()
+		; ++iter
+	) {
+		const int fd = *iter;
+
+		if (read_fd_set) {
+			FD_SET(fd, read_fd_set);
+		}
+
+		if (error_fd_set) {
+			FD_SET(fd, read_fd_set);
+		}
+
+		if (max_fd && (*max_fd < fd)) {
+			*max_fd = fd;
+		}
+	}
+
+	return 0;
+}
+
+void
+PcapManager::process(void)
+{
```

Comment:
In this PR, I cannot find where this is called. I guess maybe from `NCPInstanceBase::Process()`?